src/rules/quotemarkRule.ts

```diff
@@ -245,3 +246,91 @@ function getJSXQuotemarkPreference(
     //   If the regular pref is backtick, use double quotes instead.
     return regularQuotemarkPreference !== "`" ? regularQuotemarkPreference : '"';
 }
+
+/**
+ * Whether this node is a type constraint in a generic type.
+ * @param  node The node to check
+ * @return Whether this node is a type constraint
+ */
+function isTypeConstraint(node: ts.StringLiteral | ts.NoSubstitutionTemplateLiteral) {
+    let parent = node.parent.parent;
+
+    // If this node doesn't have a grandparent, it's not a type constraint
+    if (parent == undefined) {
+        return false;
+    }
+
+    // Iterate through all levels of union, intersection, or parethesized types
+    while (
+        parent.kind === ts.SyntaxKind.UnionType ||
+        parent.kind === ts.SyntaxKind.IntersectionType ||
+        parent.kind === ts.SyntaxKind.ParenthesizedType
+    ) {
+        parent = parent.parent;
+    }
+
+    return (
+        // If the next level is a type reference, the node is a type constraint
+        parent.kind === ts.SyntaxKind.TypeReference ||
+        // If the next level is a type parameter, the node is a type constraint
+        parent.kind === ts.SyntaxKind.TypeParameter
+    );
+}
+
+/**
+ * Whether this node is the signature of a property or method in a type.
+ * @param  node The node to check
+ * @return Whether this node is a property/method signature.
+ */
+function isSignature(node: ts.StringLiteral | ts.NoSubstitutionTemplateLiteral) {
+    let parent = node.parent;
+
+    if (ts.version < "2.7.1" && node.parent.kind === ts.SyntaxKind.LastTypeNode) {
+        // In older versions, there's a "LastTypeNode" here
+        parent = parent.parent;
+    }
+
+    return (
+        // This captures the kebab-case property names in type definitions
+        parent.kind === ts.SyntaxKind.PropertySignature ||
+        // This captures the kebab-case method names in type definitions
+        parent.kind === ts.SyntaxKind.MethodSignature
+    );
+}
+
+/**
+ * Whether this node is the method or property name in an assignment/declaration.
+ * @param  node The node to check
+ * @return Whether this node is the name in an assignment/decleration.
+ */
+function isNameInAssignment(node: ts.StringLiteral | ts.NoSubstitutionTemplateLiteral) {
+    if (
+        node.parent.kind !== ts.SyntaxKind.PropertyAssignment &&
+        node.parent.kind !== ts.SyntaxKind.MethodDeclaration
+    ) {
+        // If the node is neither a property assignment or method declaration, it's not a name in an assignment
+        return false;
+    }
+
+    return (
+        // In typescript below 2.7.1, don't change values either
+        ts.version < "2.7.1" ||
```

Comment:
Directly checking `ts.version` is a little unusual... maybe there's a cleaner way? Thoughts @adidahiya? 

Also, @ericbf, could you elaborate on why this is necessary?

Reply:
Typescript below the listed version fails compilation (or otherwise misbehaves) when backticks are used in some places where it succeeds in future versions. For instance, ``typeof str === `string``` does not narrow the type of `str` in TS versions below 2.7.1, causing compilation errors. In the other cases, older TS throws errors where NoSubstitutionLiterals are used instead of single/double quoted strings, but works fine in above 2.7.1; rather than cripple the rule to support order versions, I split it up.

Reply:
See the `typeof` issue [here](https://github.com/Microsoft/TypeScript/issues/20949). Its milestone was 2.7 (hence the split in this PR).
And [here's](https://github.com/Microsoft/TypeScript/issues/16592) the issue for backticks strings as types. This one was 2.6, but rather than doing multiple splits, I just did one at 2.7.

Reply:
Created corresponding issue, #4645

Reply:
Thanks! This LGTM but before merging I'd like to hear from adidahiya on whether there's a cleaner way to check on these things.

Reply:
I agree it's unusual and should be avoided if possible, but this seems like a reasonable use case for checking ts.version.

However, can you please use the semver package and the helper function from `parse.ts` to check version ranges instead?

```ts
import { lt } from "semver";
import { getNormalizedTypescriptVersion } from "../../verify/parse";

function hasOldTscBacktickBehavior() {
    return lt(getNormalizedTypescriptVersion(), "2.7.1");
}
```